pkg/autodiscovery/providers/kubecontainer.go

```diff
@@ -0,0 +1,315 @@
+	return &KubeContainerConfigProvider{
+		workloadmetaStore: workloadmeta.GetGlobalStore(),
+		configErrors:      make(map[string]ErrorMsgSet),
+		podCache:          make(map[string]*workloadmeta.KubernetesPod),
+		containerCache:    make(map[string]*workloadmeta.Container),
+	}, nil
+}
+
+// String returns a string representation of the KubeContainerConfigProvider
+func (k *KubeContainerConfigProvider) String() string {
+	return names.KubeContainer
+}
+
+// Collect retrieves all running pods and extract AD templates from their annotations.
+func (k *KubeContainerConfigProvider) Collect(ctx context.Context) ([]integration.Config, error) {
+	k.once.Do(func() {
+		go k.listen()
+	})
+
+	k.Lock()
+	k.upToDate = true
+	k.Unlock()
+
+	return k.generateConfigs()
+}
+
+func (k *KubeContainerConfigProvider) listen() {
+	const name = "ad-kubecontainerprovider"
+
+	k.Lock()
+	k.streaming = true
+	health := health.RegisterLiveness(name)
+	defer func() {
+		err := health.Deregister()
+		if err != nil {
+			log.Warnf("error de-registering health check: %s", err)
+		}
+	}()
+	k.Unlock()
+
+	podCh := k.workloadmetaStore.Subscribe(name, workloadmeta.NormalPriority, workloadmeta.NewFilter(
+		[]workloadmeta.Kind{workloadmeta.KindKubernetesPod},
+		workloadmeta.SourceNodeOrchestrator,
+		workloadmeta.EventTypeAll,
+	))
+
+	contCh := k.workloadmetaStore.Subscribe(name, workloadmeta.NormalPriority, workloadmeta.NewFilter(
+		[]workloadmeta.Kind{workloadmeta.KindContainer},
+		workloadmeta.SourceRuntime,
+		workloadmeta.EventTypeAll,
+	))
+
+	for {
+		select {
+		case evBundle, ok := <-podCh:
+			if !ok {
+				return
+			}
+
+			k.processOrchestratorEvents(evBundle)
+
+		case evBundle, ok := <-contCh:
+			if !ok {
+				return
+			}
+
+			k.processRuntimeEvents(evBundle)
+
+		case <-health.C:
+
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) processOrchestratorEvents(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+
+	for _, event := range evBundle.Events {
+		switch event.Type {
+		case workloadmeta.EventTypeSet:
+			pod := event.Entity.(*workloadmeta.KubernetesPod)
+			k.addPod(pod)
+		case workloadmeta.EventTypeUnset:
+			pod := event.Entity.(*workloadmeta.KubernetesPod)
+			k.deletePod(pod)
+		default:
+			log.Errorf("cannot handle event of type %d", event.Type)
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) processRuntimeEvents(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+
+	for _, event := range evBundle.Events {
+		switch event.Type {
+		case workloadmeta.EventTypeSet:
+			container := event.Entity.(*workloadmeta.Container)
+			k.addContainer(container)
+		case workloadmeta.EventTypeUnset:
+			container := event.Entity.(*workloadmeta.Container)
+			k.deleteContainer(container)
+		default:
+			log.Errorf("cannot handle event of type %d", event.Type)
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) addContainer(e *workloadmeta.Container) {
+	k.Lock()
+	defer k.Unlock()
+	// TODO do 5 second container delay thing here
```

Comment:
I thought we could get rid of the delay thing, so I went digging again to confirm. Since I already did the digging anyway, here's a summary that may include bits irrelevant to this PR:

* This seems to have been introduced to deal with short lived containers in #4048. 5 seconds sounds excessive for today's Agent, since the container config provider is polled every 1 second. This does mean that the default polling needs to be faster than 5 seconds for this provider (regardless of kubernetes or non-kubernetes environment).
* Using a `time.AfterFunc` is racy, there's no way to guarantee that one will be executed after the other if an unset, then set event occur in quick succession. You may want to explore a different solution to ensure correct ordering.
* This hack is only really necessary because config providers are not event based. I'll be working on changing this soon, so maybe don't spend a lot of time exploring a different solution to the above 😅 ?

Reply:
There could be some better solution here where the pending 'delete' operation is queryable and one could "chain" the re-addition to that (or cancel maybe?).
But given what you mentioned about changing the way this works soon, I'm going to leave the current `time.AfterFunc` approach since we know it works and it is relatively simple (despite the raceyness of it)